src/tests/mission.cpp

```diff
@@ -6,49 +6,91 @@
+void Mission::run()
+{
+    std::vector<std::shared_ptr<dcsdk::MissionItem>> items = assembleMissionItems();
+
+    uploadMission(items);
+    std::vector<std::shared_ptr<dcsdk::MissionItem>> downloaded_items = downloadMission();
+
+    compareMissions(items, downloaded_items);
+}
+
+std::vector<std::shared_ptr<dcsdk::MissionItem>> Mission::assembleMissionItems()
 {
     cout << "Number of waypoints: " << _config.num_waypoints << endl;
 
-    vector<shared_ptr<MissionItem>> mission_items;
+    std::vector<std::shared_ptr<dcsdk::MissionItem>> items{};
 
     for (int i = 0; i < _config.num_waypoints; ++i) {
         float altitude = 10.f + (float)i;
         double latitude = 47.398170327054473 + (double)i * 1e-5;
-        mission_items.push_back(makeMissionItem(latitude, 8.5456490218639658, altitude));
+        items.push_back(makeMissionItem(latitude, 8.5456490218639658, altitude));
     }
 
+    return items;
+}
+
+void Mission::uploadMission(const std::vector<std::shared_ptr<dcsdk::MissionItem>>& items)
+{
     cout << "Uploading mission..." << endl;
-    auto prom = make_shared<promise<Mission::Result>>();
-    auto future_result = prom->get_future();
-    _mission.upload_mission_async(mission_items, [prom](Mission::Result result) { prom->set_value(result); });
+    promise<dcsdk::Mission::Result> prom{};
+    auto fut = prom.get_future();
+
+    _mission.upload_mission_async(
+        items, [&prom](dcsdk::Mission::Result result) { prom.set_value(result); });
 
     // wait until uploaded
-    const Mission::Result result = future_result.get();
-    if (result != Mission::Result::SUCCESS) {
-        cout << "Mission upload failed (" << Mission::result_str(result) << "), exiting." << endl;
-        return Result::Failed;
-    }
-    cout << "Mission uploaded." << endl;
+    const dcsdk::Mission::Result result = fut.get();
+    EXPECT_EQ(result, dcsdk::Mission::Result::SUCCESS);
+}
 
-    return Result::Success;
+std::vector<std::shared_ptr<dcsdk::MissionItem>> Mission::downloadMission()
+{
+    cout << "Downloading mission..." << endl;
+    promise<pair<dcsdk::Mission::Result, std::vector<std::shared_ptr<dcsdk::MissionItem>>>> prom{};
+    auto fut = prom.get_future();
+
+    _mission.download_mission_async(
+        [&prom](dcsdk::Mission::Result result,
+                std::vector<std::shared_ptr<dcsdk::MissionItem>> items) {
+            prom.set_value(make_pair(result, items));
+        });
+
+    auto value = fut.get();
+    const dcsdk::Mission::Result& result = value.first;
+    const std::vector<std::shared_ptr<dcsdk::MissionItem>>& items = value.second;
+
+    // wait until uploaded
+    EXPECT_EQ(result, dcsdk::Mission::Result::SUCCESS);
+
+    return items;
+}
+
+void Mission::compareMissions(const std::vector<std::shared_ptr<dcsdk::MissionItem>>& items_a,
+                              const std::vector<std::shared_ptr<dcsdk::MissionItem>>& items_b)
+{
+    EXPECT_EQ(items_a.size(), items_b.size());
```

Comment:
This needs to be an assertion. Otherwise you can access invalid data.

Reply:
True.

Reply:
Fixed in d27d0f8.